openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb

```diff
@@ -0,0 +1,744 @@
+    "    print(\"validation logged metrics\", trainer.logged_metrics)\n",
+    "\n",
+    "    val_loss = trainer.logged_metrics[\"Discriminator val loss\"]\n",
+    "\n",
+    "    return {\"val_loss\": val_loss}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "derived-bride",
+   "metadata": {},
+   "source": [
+    "## Time to start a federated learning experiment"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "mature-renewal",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# create an experimnet in federation\n",
+    "experiment_name = \"PL_MNIST_test_experiment\"\n",
+    "fl_experiment = FLExperiment(federation=federation, experiment_name=experiment_name)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "lightweight-causing",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "fl_experiment.start(\n",
+    "    model_provider=MI,\n",
+    "    task_keeper=TI,\n",
+    "    data_loader=fed_dataset,\n",
+    "    rounds_to_train=5,\n",
+    "    opt_treatment=\"CONTINUE_GLOBAL\",\n",
+    "    device_assignment_policy=\"CUDA_PREFERRED\",\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f1543a36",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "fl_experiment.stream_metrics()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f987d2c4",
+   "metadata": {},
+   "source": [
+    "## Check the images generated by the model in tensorboard"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "fec7a708",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install -r ../envoy/sd_requirements.txt\n",
+    "import sys\n",
+    "\n",
+    "sys.path.insert(1, \"../envoy\")\n",
+    "from mnist_shard_descriptor import MnistShardDescriptor"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "2bc77cd8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "fed_dataset = MnistFedDataset(train_bs=256, valid_bs=64)\n",
+    "fed_dataset.shard_descriptor = MnistShardDescriptor(rank_worldsize=\"1,1\")\n",
+    "\n",
+    "last_model = fl_experiment.get_last_model()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5f821972",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "val_imgs, _ = next(iter(fed_dataset.get_valid_loader()))\n",
+    "\n",
+    "z = torch.randn(val_imgs.shape[0], 100)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9549d1ab",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "last_model.train_generator(val_imgs, z, logdir=\"generated_images\")\n",
+    "pass"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5f5c09f8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# View it in tensorboard\n",
+    "\n",
+    "# %load_ext tensorboard\n",
```

Comment:
I believe we only need 1 way to launch TensorBoard, please double-check and remove one of this